src/rust/engine/src/intrinsics.rs

```diff
@@ -1,58 +1,131 @@
 use crate::context::Context;
-use crate::core::{throw, TypeId, Value};
+use crate::core::{throw, Value};
 use crate::externs;
 use crate::nodes::MultiPlatformExecuteProcess;
 use crate::nodes::{lift_digest, DownloadedFile, NodeFuture, Snapshot};
+use crate::tasks::Intrinsic;
+use crate::types::Types;
+
 use boxfuture::Boxable;
 use bytes;
 use futures::future::{self as future03, TryFutureExt};
 use futures01::{future, Future};
 use hashing;
+use indexmap::IndexMap;
+
 use std::path::PathBuf;
 
-pub fn run_intrinsic(
-  input: TypeId,
-  product: TypeId,
-  context: Context,
-  value: Value,
-) -> NodeFuture<Value> {
-  let types = &context.core.types;
-  if product == types.process_result && input == types.multi_platform_process_request {
-    multi_platform_process_request_to_process_result(context, value)
-  } else if product == types.files_content && input == types.directory_digest {
-    directory_digest_to_files_content(context, value)
-  } else if product == types.directory_digest && input == types.directory_with_prefix_to_strip {
-    directory_with_prefix_to_strip_to_digest(context, value)
-  } else if product == types.directory_digest && input == types.directory_with_prefix_to_add {
-    directory_with_prefix_to_add_to_digest(context, value)
-  } else if product == types.snapshot && input == types.directory_digest {
-    digest_to_snapshot(context, value)
-  } else if product == types.directory_digest && input == types.directories_to_merge {
-    directories_to_merge_to_digest(context, value)
-  } else if product == types.snapshot && input == types.url_to_fetch {
-    url_to_fetch_to_snapshot(context, value)
-  } else if product == types.snapshot && input == types.path_globs {
-    path_globs_to_snapshot(context, value)
-  } else if product == types.directory_digest && input == types.input_files_content {
-    input_files_content_to_digest(context, value)
-  } else if product == types.snapshot && input == types.snapshot_subset {
-    snapshot_subset_to_snapshot(context, value)
-  } else {
-    panic!("Unrecognized intrinsic: {:?} -> {:?}", input, product)
+type IntrinsicFn = Box<dyn Fn(Context, Vec<Value>) -> NodeFuture<Value> + Send + Sync + 'static>;
+
+pub struct Intrinsics {
+  intrinsics: IndexMap<Intrinsic, IntrinsicFn>,
+}
+
+impl Intrinsics {
+  pub fn new(types: &Types) -> Intrinsics {
+    let mut intrinsics: IndexMap<Intrinsic, IntrinsicFn> = IndexMap::new();
+    intrinsics.insert(
+      Intrinsic {
+        product: types.directory_digest,
+        inputs: vec![types.input_files_content],
```

Comment:
It would be nice to be able to tie the inputs length to the length on the values vec expected by the intrinsic function. The &args[0], &args[1] look scary in isolation (what if there are 3 input types and the function evolves to only using 2 or needing 4. There is and will be a small enough set of intrinsics that just making sure these align by hand is probably good enough.

Reply:
> There is and will be a small enough set of intrinsics that just making sure these align by hand is probably good enough.

Yea. For better or worse, these have evolved away from a large switch statement and toward dynamic dispatch like this, because it makes [the callsite](https://github.com/pantsbuild/pants/blob/87291398b4dd7f622abb88d829608c23209011ab/src/rust/engine/src/nodes.rs#L175-L179) clearer. It's possible that moving back toward the switch statement is feasible, but for now I think that the fact these are now isolated to one file makes this lack of type safety manageable.